pil-analyzer/src/type_inference.rs

```diff
@@ -420,7 +425,18 @@ impl<'a> TypeChecker<'a> {
         expressions: &mut [(&mut Expression, ExpectedType)],
     ) -> Result<(), String> {
         for (e, expected_type) in expressions {
-            self.expect_type_with_flexibility(expected_type, e)?;
+            let result = self.expect_type_with_flexibility(expected_type, e);
```

Comment:
I think it would be better to do this change inside `expect_type_with_flexibility`.

As I said in other comments: Once type checking fails, you cannot re-check the same expression with a different expected type because it has run unification which adds substitutions and trait constraints on types and so on. If you then re-run it with a different expectation, these substitutions are still there.